lib/handler/configurator/proxy.c

```diff
@@ -362,4 +399,13 @@ void h2o_proxy_register_configurator(h2o_globalconf_t *conf)
     h2o_configurator_define_command(&c->super, "proxy.emit-x-forwarded-headers",
                                     H2O_CONFIGURATOR_FLAG_GLOBAL | H2O_CONFIGURATOR_FLAG_EXPECT_SCALAR,
                                     on_config_emit_x_forwarded_headers);
+#define DEFINE_CMD(name, cb)                                                                                                       \
+    h2o_configurator_define_command(&c->super, name, H2O_CONFIGURATOR_FLAG_ALL_LEVELS | H2O_CONFIGURATOR_FLAG_EXPECT_SCALAR, cb)
+    DEFINE_CMD("proxy.header.add", on_config_header_add);
+    DEFINE_CMD("proxy.header.append", on_config_header_append);
+    DEFINE_CMD("proxy.header.merge", on_config_header_merge);
+    DEFINE_CMD("proxy.header.set", on_config_header_set);
+    DEFINE_CMD("proxy.header.setifempty", on_config_header_setifempty);
+    DEFINE_CMD("proxy.header.unset", on_config_header_unset);
+#undef DEFINE_CMD
```

Comment:
How about creating a function that adds various favors (e.g. `add`, `append`, ...) in lib/handler/configurator/util.c, instead of adding `h2o_on_config_header_2arg` and using macro to create variations in each source file.

E.g., How aboutcreating a function that looks something like `h2o_configurator_define_headers_commands(h2o_configurator_t *conf, const char *prefix, h2o_headers_commands_t *(*get_commands)(h2o_configurator_t *))`, and call it here like:

```c
h2o_configurator_define_headers_comands(&c->super, "proxy.header", get_commands);
```

The invocation would define all the `proxy.header.*` directives.

`get_commands` function could be defined as:
```c
static h2o_headers_commands_t *get_commands(h2o_configurator_t *conf)
{
    struct proxy_configurator_t *self = (void *)cmd->configurator;
    return &self->vars->header_cmds;
}
```

The merits of using this approach would be that the code will become less redundant, and that new directives that modifies the header modification commands can be added just be changing one place.